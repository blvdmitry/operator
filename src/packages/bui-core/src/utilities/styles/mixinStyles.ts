import type * as T from "../../types";
import responsiveStyles, { type ResponsiveStyles } from "./responsiveStyles";
import { toKebabCase } from "../helpers";

const mixinStyles = (mixin?: T.Mixin): ResponsiveStyles => {
  if (!mixin) return {};

  const mixinKeys = Object.keys(mixin) as Array<keyof T.Mixin>;

  const styles = mixinKeys.reduce(
    (res: ResponsiveStyles, key: keyof T.Mixin) => {
      const value = mixin[key];

      return {
        ...res,
        ...responsiveStyles(value, "mixin", toKebabCase(key)),
      };
    },
    {}
  );

  return styles;
};

export default mixinStyles;
