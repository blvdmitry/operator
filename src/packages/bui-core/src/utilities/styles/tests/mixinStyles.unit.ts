import mixinStyles from "../mixinStyles";

describe("styles/mixinStyles", () => {
  it("resolves mixin", () => {
    const result = mixinStyles({
      height: {
        s: 4,
        l: 8,
      },
      position: "relative",
      insetBlockStart: 20,
    });

    expect(result).toMatchSnapshot();
  });
});
