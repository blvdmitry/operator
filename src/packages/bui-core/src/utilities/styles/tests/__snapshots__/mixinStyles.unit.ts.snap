// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styles/mixinStyles resolves mixin 1`] = `
{
  "--bui_mixin_position--s": "relative",
  "--bui_mixin_spaced_height--l": 8,
  "--bui_mixin_spaced_height--s": 4,
  "--bui_mixin_spaced_inset-block-start--s": 20,
}
`;
